src/x/debug/triggering_profile.go

```diff
@@ -0,0 +1,212 @@
+// THE SOFTWARE.
+
+package debug
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"runtime/pprof"
+	"text/template"
+	"time"
+
+	"github.com/m3db/m3/src/x/instrument"
+
+	"go.uber.org/zap"
+)
+
+const (
+	// ContinuousCPUProfileName is the name of continuous CPU profile.
+	ContinuousCPUProfileName = "cpu"
+)
+
+var (
+	defaultConditional     = func() bool { return true }
+	defaultInterval        = time.Second
+	errNoFilePathTemplate  = errors.New("no file path template")
+	errProfileName         = errors.New("no profile name")
+	errNoInstrumentOptions = errors.New("no instrument options")
+	errAlreadyOpen         = errors.New("already open")
+	errNotOpen             = errors.New("not open")
+)
+
+// ContinuousFileProfile is a profile that runs continously
+// to a file using a template for a file name.
+type ContinuousFileProfile struct {
+	filePathTemplate *template.Template
+	profileName      string
+	profileDuration  time.Duration
+	profileDebug     int
+	conditional      func() bool
+	interval         time.Duration
+
+	logger *zap.Logger
+
+	closeCh chan struct{}
+}
+
+// ContinuousFileProfileOptions is a set of continuous file profile options.
+type ContinuousFileProfileOptions struct {
+	FilePathTemplate  string
+	ProfileName       string
+	ProfileDuration   time.Duration
+	ProfileDebug      int
+	Conditional       func() bool
+	Interval          time.Duration
+	InstrumentOptions instrument.Options
+}
+
+// ContinuousFileProfilePathParams is the params used to construct
+// a file path.
+type ContinuousFileProfilePathParams struct {
+	ProfileName string
+	UnixTime    int64
+}
+
+// NewContinuousFileProfile returns a new continuous file profile.
+func NewContinuousFileProfile(
+	opts ContinuousFileProfileOptions,
+) (*ContinuousFileProfile, error) {
+	if opts.FilePathTemplate == "" {
+		return nil, errNoFilePathTemplate
+	}
+	if opts.ProfileName == "" {
+		return nil, errProfileName
+	}
+	if opts.Conditional == nil {
+		opts.Conditional = defaultConditional
+	}
+	if opts.Interval == 0 {
+		opts.Interval = defaultInterval
+	}
+	if opts.InstrumentOptions == nil {
+		return nil, errNoInstrumentOptions
+	}
+
+	tmpl, err := template.New("fileName").Parse(opts.FilePathTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ContinuousFileProfile{
+		filePathTemplate: tmpl,
+		profileName:      opts.ProfileName,
+		profileDuration:  opts.ProfileDuration,
+		profileDebug:     opts.ProfileDebug,
+		conditional:      opts.Conditional,
+		interval:         opts.Interval,
+		logger:           opts.InstrumentOptions.Logger(),
+	}, nil
+}
+
+// Start will start the continuous file profile.
+func (c *ContinuousFileProfile) Start() error {
+	if c.closeCh != nil {
+		return errAlreadyOpen
+	}
+
+	c.closeCh = make(chan struct{})
+	go c.run()
+	return nil
+}
+
+// Stop will stop the continuous file profile.
+func (c *ContinuousFileProfile) Stop() error {
+	if c.closeCh == nil {
```

Comment:
Does this need a lock for race of multiple start/stop calls?

Reply:
Only mean to be use in a single threaded way, so no problem here.